consolidate printer OutputOpts w PrintOpts

juanvallejo opened the pull request:
This patch removes the use of printers.OutputOptions in favor of only
having a single struct for setting / passing printer options set by user
flags.

This PR is related to https://github.com/kubernetes/kubernetes/pull/50113 and aims to break the set of changes 
introduced in [this commit](https://github.com/kubernetes/kubernetes/pull/50113/commits/f4d717450f431c8a9d5265a7c06d33b26bb5132b) in order to make them easier to review.

**Release note**:
```release-note
NONE
```

cc @fabianofranz @mengqiy @shiywang @seans3

juanvallejo committed:
consolidate printer OutputOpts w PrintOpts

This patch removes the use of printers.OutputOptions in favor of only
having a single struct for setting / passing printer options set by user
flags.

juanvallejo commented on pkg/kubectl/cmd/util/factory_builder.go:
```diff
@@ -48,7 +48,7 @@ func NewBuilderFactory(clientAccessFactory ClientAccessFactory, objectMappingFac
 }
 
-func (f *ring2Factory) PrinterForCommand(cmd *cobra.Command, isLocal bool, options printers.PrintOptions) (printers.ResourcePrinter, error) {
+func (f *ring2Factory) PrinterForCommand(isLocal bool, options printers.PrintOptions) (printers.ResourcePrinter, error) {
```
@fabianofranz given that this method would no longer depend on `*cobra.Command` with the proposed changes, I think it would make sense to rename this method or combine it with another printing method on the factory. WDYT?

dims commented:
/ok-to-test

fabianofranz commented:
PTAL @smarterclayton regarding your comment in another PR.

smarterclayton commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -171,29 +178,21 @@ func extractOutputOptions(cmd *cobra.Command) *printers.OutputOptions {
-		AllowMissingKeys: allowMissingTemplateKeys,
-	}
-}
+	options.FmtType = outputFormat
```
Rename this to be full names - `OutputFormatType` and `OutputFormatArgument`

smarterclayton replied:
Also, move these to the top of the PrintOptions struct

smarterclayton commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -235,13 +233,12 @@ type BuilderFactory interface {
 	// Returns a printer, true if the printer is generic (is not internal), or
 	// an error if a printer could not be found.
-	PrinterForMapping(cmd *cobra.Command, isLocal bool, outputOpts *printers.OutputOptions, mapping *meta.RESTMapping, withNamespace bool) (printers.ResourcePrinter, error)
+	PrinterForMapping(isLocal bool, options printers.PrintOptions, mapping *meta.RESTMapping, withNamespace bool) (printers.ResourcePrinter, error)
```
Nuke local while you are in here.

smarterclayton replied:
Why isn't WithNamespace on PrintOptions as well?

smarterclayton commented:
+1 to rename - probably `PrinterForOptions` and move the PrintOptions to the first argument.  Please ping me when comments addressed.

juanvallejo commented:
@smarterclayton thanks, comments addressed: 
- nuked the `isLocal` arg from factory_builder printer methods (replaced previous mapper logic with `mapper, typer = f.objectMappingFactory.Object()` from https://github.com/kubernetes/kubernetes/pull/55660)
- renamed PrintOptions output format fields
- moved `withNamespace` arg to be specified through PrintOptions
- renamed `PrinterForCommand` -> `PrinterForOptions`

smarterclayton commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -327,14 +328,14 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
 			}
 
-			var outputOpts *printers.OutputOptions
+			printOpts := *cmdutil.ExtractCmdPrintOptions(cmd, printWithNamespace)
```
It would be better if ExtractCmdPrintOptions returned a pointer.  This is generally considered a code smell (to have to `*` a function return).

smarterclayton commented on pkg/kubectl/cmd/testing/fake.go:
```diff
@@ -351,17 +351,17 @@ func (f *FakeFactory) Describer(*meta.RESTMapping) (printers.Describer, error) {
 }
 
-func (f *FakeFactory) PrinterForCommand(cmd *cobra.Command, isLocal bool, outputOpts *printers.OutputOptions, options printers.PrintOptions) (printers.ResourcePrinter, error) {
+func (f *FakeFactory) PrinterForOptions(options printers.PrintOptions) (printers.ResourcePrinter, error) {
```
make this a pointer

smarterclayton commented on pkg/kubectl/cmd/testing/fake.go:
```diff
@@ -489,7 +489,7 @@ func (f *FakeFactory) PrintObject(cmd *cobra.Command, isLocal bool, mapper meta.
 }
 
-func (f *FakeFactory) PrinterForMapping(cmd *cobra.Command, isLocal bool, outputOpts *printers.OutputOptions, mapping *meta.RESTMapping, withNamespace bool) (printers.ResourcePrinter, error) {
+func (f *FakeFactory) PrinterForMapping(printOpts printers.PrintOptions, mapping *meta.RESTMapping) (printers.ResourcePrinter, error) {
```
make this a pointer

smarterclayton commented:
Two more small changes please.

juanvallejo committed:
pass printOpts by pointer

smarterclayton commented:
Wonderful

/lgtm
/approve no-issue

smarterclayton commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -106,23 +106,11 @@ func ValidateOutputArgs(cmd *cobra.Command) error {
 }
 
-// PrinterForCommand returns the printer for the outputOptions (if given) or
+// printerForOptions returns the printer for the outputOptions (if given) or
```
Nit, can fix later

smarterclayton commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -662,43 +664,44 @@ func cmdSpecifiesOutputFmt(cmd *cobra.Command) bool {
 
 // outputOptsForMappingFromOpenAPI looks for the output format metatadata in the
 // openapi schema and returns the output options for the mapping if found.
```
Godoc should say "modifies the passed print options" and the name of the method should reflect it - `updatePrintOptionsForOpenAPI`

juanvallejo commented:
@smarterclayton removing the `local` argument from `f.PrinterForOptions` / `f.PrinterForMapping` methods causes this PR to depend on https://github.com/kubernetes/kubernetes/pull/55660 in order for the last few tests to pass. Since I am now [initializing the mapper in `PrinterForOptions` to `f.objectMappingFactory.Object`](https://github.com/kubernetes/kubernetes/pull/54811/files#diff-a9b5082d3df09126fbcd118712148907R54), printing things such as TPRs with the `-o name` option won't work correctly until the changes from pull/55660 go in.

juanvallejo commented:
/test pull-kubernetes-e2e-kops-aws

juanvallejo commented:
@smarterclayton wondering if you could re-tag this once more? thx!

smarterclayton commented:
/lgtm

smarterclayton commented:
Approved, cleanup necessary for 1.10 changes
